packages/core/src/browser/xhrProxy.ts

```diff
@@ -69,85 +80,82 @@ export function resetXhrProxy() {
-            return
-          }
-          hasBeenReported = true
-
-          const xhrCompleteContext: XhrCompleteContext = {
-            ...xhrPendingContext,
-            duration: elapsed(xhrPendingContext.startClocks.timeStamp, timeStampNow()),
-            response: this.response as string | undefined,
-            status: this.status,
-          }
-
-          onRequestCompleteCallbacks.forEach((callback) => callback(xhrCompleteContext))
-        }
-
-        this.addEventListener('loadend', monitor(reportXhr))
-
-        beforeSendCallbacks.forEach((callback) => callback(xhrPendingContext, this))
+    this._datadog_xhr.startTime = relativeNow()
+    this._datadog_xhr.startClocks = clocksNow()
+    this._datadog_xhr.isAborted = false
+
+    let hasBeenReported = false
+    const originalOnreadystatechange = this.onreadystatechange
+    const onreadystatechange = function (this: BrowserXHR<XhrStartContext>) {
+      if (this.readyState === XMLHttpRequest.DONE) {
+        // Try to report the XHR as soon as possible, because the XHR may be mutated by the
+        // application during a future event. For example, Angular is calling .abort() on
+        // completed requests during a onreadystatechange event, so the status becomes '0'
+        // before the request is collected.
+        onEnd()
       }
-    })
 
-    return originalXhrSend.apply(this, arguments as any)
-  }
+      if (originalOnreadystatechange) {
+        originalOnreadystatechange.apply(this, arguments as any)
+      }
+    }
 
-  XMLHttpRequest.prototype.abort = function (this: BrowserXHR) {
-    callMonitored(() => {
-      if (this._datadog_xhr) {
-        this._datadog_xhr.isAborted = true
+    const onEnd = monitor(() => {
+      this.removeEventListener('loadend', onEnd)
+      // if the onreadystatechange hasn't been overridden by the user after the send()
+      if (this.onreadystatechange === onreadystatechange) {
+        this.onreadystatechange = originalOnreadystatechange
       }
+      if (hasBeenReported) {
+        return
+      }
+      hasBeenReported = true
+      reportXhr(this, this._datadog_xhr!)
     })
-    return originalXhrAbort.apply(this, arguments as any)
+    this.onreadystatechange = onreadystatechange
+    this.addEventListener('loadend', onEnd)
+
+    beforeSendCallbacks.forEach((callback) => callback(this._datadog_xhr!, this))
+  })
+
+  return originalXhrSend.apply(this, arguments as any)
+}
+
+function abortXhr(this: BrowserXHR<XhrStartContext>) {
+  callMonitored(() => {
+    if (this._datadog_xhr) {
+      this._datadog_xhr.isAborted = true
+    }
+  })
+  return originalXhrAbort.apply(this, arguments as any)
+}
+
+function reportXhr(xhr: BrowserXHR<XhrStartContext>, pendingContext: XhrStartContext) {
+  const xhrCompleteContext: XhrCompleteContext = {
+    ...pendingContext,
+    duration: elapsed(pendingContext.startClocks.timeStamp, timeStampNow()),
+    response: xhr.response as string | undefined,
+    status: xhr.status,
   }
```

Comment:
what about reinforcing the design by having something like:
```suggestion
function reportXhr(xhr: BrowserXHR<XhrCompleteContext>) {
  if (xhr._datadog_xhr) {
    xhr._datadog_xhr.duration = elapsed(pendingContext.startClocks.timeStamp, timeStampNow())
    ...
  }
```

Reply:
Very good point. Seems much nicer !

Reply:
it seems good enough to me